cdap-client/src/main/java/co/cask/cdap/client/AuthorizationClient.java

```diff
@@ -101,22 +110,106 @@ public void revoke(EntityId entity) throws IOException, UnauthenticatedException
+                                                principal.getName(), role.getName()));
+    HttpRequest request = HttpRequest.put(url).build();
+    executeRoleRequest(role, request);
+  }
+
+  public void removeRoleFromPrincipal(Role role, Principal principal) throws IOException, FeatureDisabledException,
+    UnauthenticatedException, UnauthorizedException, RoleNotFoundException {
+    URL url = config.resolveURLV3(String.format(AUTHORIZATION_BASE + "%s/%s/roles/%s", principal.getType(),
+                                                principal.getName(), role.getName()));
+    HttpRequest request = HttpRequest.delete(url).build();
+    executeRoleRequest(role, request);
+  }
+
+  private Set<Role> listRolesHelper(@Nullable Principal principal) throws IOException, FeatureDisabledException,
+    UnauthenticatedException, UnauthorizedException {
+    URL url = principal == null ? config.resolveURLV3(AUTHORIZATION_BASE + "roles") :
+      config.resolveURLV3(String.format(AUTHORIZATION_BASE + "%s/%s/roles", principal.getType(), principal.getName()));
+    HttpRequest request = HttpRequest.get(url).build();
+    HttpResponse response = executeRequest(request, HttpURLConnection.HTTP_FORBIDDEN,
+                                           HttpURLConnection.HTTP_NOT_IMPLEMENTED);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_OK) {
+      return ObjectResponse.fromJsonBody(response, new TypeToken<Set<Role>>() {
+      }).getResponseObject();
+    }
+    throw new IOException(String.format("Cannot list roles. Reason: %s", response.getResponseBodyAsString()));
+  }
+
+  private void executeRoleRequest(Role role, HttpRequest request) throws IOException, UnauthenticatedException,
+    FeatureDisabledException, UnauthorizedException, RoleNotFoundException {
+    HttpResponse httpResponse = executeRequest(request, HttpURLConnection.HTTP_FORBIDDEN,
+                                               HttpURLConnection.HTTP_NOT_IMPLEMENTED,
+                                               HttpURLConnection.HTTP_NOT_FOUND);
+    if (httpResponse.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new RoleNotFoundException(role);
+    }
+  }
+
+  private HttpResponse executeRequest(HttpRequest request, int... allowedErrorCodes)
     throws IOException, UnauthenticatedException, FeatureDisabledException, UnauthorizedException {
-    HttpResponse response = restClient.execute(request, config.getAccessToken(), HttpURLConnection.HTTP_FORBIDDEN,
-                                               HttpURLConnection.HTTP_NOT_IMPLEMENTED);
+    HttpResponse response = restClient.execute(request, config.getAccessToken(), allowedErrorCodes);
     if (HttpURLConnection.HTTP_FORBIDDEN == response.getResponseCode()) {
-      throw new UnauthorizedException(principal, action, entity);
+      // TODO:(CDAP-5302) Include the logged in username here
+      throw new UnauthorizedException(String.format("Unauthorized to perform %s %s with body %s",
```

Comment:
the message here is strange and could be really long. when you get the logged in user, will you be able to use the other constructor for ``UnauthorizedException``?